### ISSUE TITLE
Configurable retention period for messages.

### ISSUE THREAD

**Issue by kdurrance** [gotify/server#376]
Pretty self explanatory. Option to automatically prune old messages from the gotify.db en masse. I have a particular chatty application for notifications, which over time could be an issue. Would be great to have configurable retention period for messages, either as a global setting, or an a per-application basis.

**Comment by jmattheis** [comment]
I'd be happy to accept a pr for this.

**Comment by StewartThomson** [comment]
https://godoc.org/github.com/robfig/cron could be useful for this

**Comment by StewartThomson** [comment]
https://godoc.org/github.com/robfig/cron could be useful for this, if a bit overkill

**Comment by IvanMazzoli** [comment]
+1 on this feature, would be of great use for me

**Comment by felixq** [comment]
+1. My app can generate around 100 messages per hour and they are only valid for less than 5 minutes.

**Comment by mateuscelio** [comment]
@jmattheis, I could work on this if you think that this feature is required.

**Comment by jmattheis** [comment]
@mateuscelio I don't think this feature is required, but feel free to work on it.

For the implementation: Each application should get a separate setting for the
retention period. The pruning should be done in a separate goroutine that
executes an sql query every minute to prune messages outside of the retention
period. The default retention period should be unlimited.

**Comment by disconn3ct** [comment]
> @mateuscelio I don't think this feature is required, but feel free to work on it.
> 
> For the implementation: Each application should get a separate setting for the retention period. The pruning should be done in a separate goroutine that executes an sql query every minute to prune messages outside of the retention period. The default retention period should be unlimited.

What about making message expiration part of the message itself? If it isn't sent, the application default is applied (null or a timestamp)

Most of my cleanup needs can be handled app by app, but there are some things that would be better with individual message expiration. ("Garage door ajar" is unlimited but "Fresh coffee downstairs" should expire in an hour.)

**Comment by eternal-flame-AD** [comment]
Maybe we can have a default TTL on the app and apps can optionally override that by sending their own per message?

**Comment by kaba72** [comment]
+1 - auto-expiry/retention period would be great to have to keep things tidy without additional add-on work

**Comment by Equilibrier** [comment]
Has this already been implemented ?

I implemented my own broker which does this, by this configuration, for example:

gotify_url: "http://example.gotify.com:3002"
client_token: "blahblah123e"
db_path: "./gotify.db"

retention:
  - app_token: "kdlasdiiodask"
    keep_last_n: 1 # unique messages per title (not per app)
    apply_per_message: true
    max_age_sec: 1000
    description: "global-events app"

and it works greatly, but unfortunately, as gotify has a bug (websocket delete message not sent to clients if you deleted from RESP API) I still have to hit F5 in the browser to see the results.

**Comment by jmattheis** [comment]
> Has this already been implemented ?

No. If so, this ticket would be closed.

**Comment by BerserkeR-Git** [comment]
I would like to see this as well on a per-application basis. Preferably as a field in the create application/edit application window, that accepts e.g. `40s`, `15m`, `26h`, `14d` for seconds/minutes/hours/days and I think it's best to have it on a rolling basis so every notification gets to live for that long instead of a hard nuke which could potentially remove a really short lived notification of that application.

Maybe there's also a possibility where you could incorporate all of these in a field `d`,`h`,`m`,`s` like `5d2h22m6s` for 5 days, 2 hours, 22 minutes and 6 seconds, just as explanation for the format, and you could choose any one of them on their own.

**Comment by Krojack** [comment]
Just made something where I could use this feature. I have alerts set at set times each 1hr or 20 past even hours. These alerts I would like to be auto removed something like 10 or so minutes else they keep piling up during the day.

I was looking to automate this using the API but I keep getting error 401 with the following:

`curl -X GET "http://192.168.0.10:8000/stream" -H "accept: application/json"`

That seems to be correct according to https://gotify.github.io/api-docs/#/message/streamMessages

**Comment by jmattheis** [comment]
@Krojack You need to provide authentication to access this api, furthermore this api is a websocket api and cannot be queried with curl. See https://github.com/gotify/server/issues/255 for other cli tools that support this. If you have any further problems, create a new ticket.

**Comment by GAS85** [comment]
@Krojack why you don not try this one https://github.com/gotify/server/issues/376#issuecomment-942328343?

**Comment by zero77** [comment]
This also works:

`sqlite3 gotify.db "DELETE FROM messages WHERE date < datetime('now', '-1 day');"`

**Comment by GableTheSnail** [comment]
Chiming in on this one due to #889 and [disconn3c nailed it](https://github.com/gotify/server/issues/376#issuecomment-2480584658).

It would be nice to even implement an expiry time at the level of the Gotify server itself so the database does not get full to the brim with messages when most of them are just "timely notifications" and not really long-term data. Perhaps a default and user-configurable expiry time like 1 month should be sufficient to expunge data that is no longer relevant?

**Comment by disconn3ct** [comment]
This feature is the main reason I (regrettably) switched to ntfy just before their huge slop import. I got tired of configuring expiration cron jobs every time I added an app.

(Regrettable because now I'm switching back and fighting cron jobs and such again.)

Added to stay on topic:
I still think the idea of hierarchical defaults makes a ton of sense. The individual message setting overrides the application setting overrides the global default. That does leave a question though. 
If I set an application to 1 day, can an individual message set itself to 1 week? Or can it only go lower? I would love to be able to set that at each level (global/app) so that, for example, an info notification ("health ok") expires in a day, but extraordinary ones ("battery on fire") expire in a month, or never.

I propose this:
- Resolution of 1 minute, but no accuracy guarantees
- Everything is optional. Unset, it acts like now (no age, no purge.)
- Global default is set in the server config.
  - No messages ever go longer. Equivalent to the [sql command earlier](https://github.com/gotify/server/issues/376#issuecomment-3337701335).
- Overrides as above:
  - message over app over global
  - each level's max caps the lifetime.
  - Concrete example: 
    - Global max: 1 year.
    - Global default: 3 months.
    - App max: 1 month, no default
  - Send:
    - Message 1 lifetime: 2 months
    - Message 2 lifetime: none
  - Message 1: 1 month
    - 2 months is dropped to 1 month by the app max, and since it is less than the global max it remains at 1 month.
  - Message 2: 3 months
    - There is no message lifetime and no app default, so there is no comparison to max at the app level
    - Lifetime is defaulted to global
- App settings are managed alongside default priority. "Default message lifetime" "Maximum message lifetime"
- Per-message setting would probably fit as an extra
  -  ``server::lifetime::[seconds|timestamp]`` maybe

**Comment by jmattheis** [comment]
I agree with most of it, but I rather have a app default than a app max. I mean if the user wants to configure a app default of 1 month, and then have some messages that are >1month, then it should be fine.

The global max should be absolute, so it's not possible to have a higher value than defined there. So It's basically this order.

* global max (can be infinite)
* global default (if not set, it's global max)
* app default (if not set, it's global default)
* message expiry (if not set, it's app default)

> Per-message setting would probably fit as an extra
  server::lifetime::[seconds|timestamp]

I'd make this a non extra field, as extras are currently only used for client side stuff. The only thing I'm not sure we store the absolute time when the message expires, or the the message lifetime. E.g.

```
{
    "message": "notification",
    "expiresAt": "2026-03-20T19:24:55.233Z",
```
or
```
{
    "message": "notification",
    "createdAt": "2026-03-17T19:24:55.233Z",
    "expiresIn": "72h"
```
I'm leaning towards the first, as it's more easy to query and to understand. We can probably support both when `POST`ing a `/message`.

This feature is on my todo list, and I'll work on it after #433.
